### PR TITLE
CBL-3309: Collection change listener objc 

### PR DESCRIPTION
**Pull request by jayahariv**
* Add delegate to listener token class and CBLDatabase/CBLCollection/CBLReplicator/CBLQuery classes now implements it.
* now removeToken function does calls the removeChangeListener callback.

### CODE REVIEW

File: Objective-C/CBLCollection.mm [hunk 1]
```diff
@@ -244,4 +263,69 @@ - (BOOL) collectionIsValid: (NSError**)error {
...
+                [docIDs addObject: docID];
+            }
+            c4dbobs_releaseChanges(changes, obs.numChanges);
+        } while(obs.numChanges > 0);
+    }
+}
+
+- (void) removeToken: (id)token {
+    CBL_LOCK(self) {
+        if (![self collectionIsValid: nil])
```
Comment (pasin) [inline Objective-C/CBLCollection.mm:318]: I think this check might not be needed. Will it cause any problems if remove the check?

File: Swift/Collection.swift [hunk 2]
```diff
@@ -244,7 +244,7 @@ public final class Collection : CollectionChangeObservable, Indexable {
     /// To remove the listener, call remove() function on the returned listener token
     ///.
     /// If the collection is deleted or the database is closed, a warning message will be logged.
-    public func addChangeListener(listener: @escaping (CollectionChange) -> Void) -> ListenerToken {
+    public func addChangeListener(listener: @escaping (CollectionChange) -> Void) -> ListenerToken? {
```
Comment (pasin) [inline Swift/Collection.swift:247]: The returned ListenerToken is non-null.

File: Swift/Collection.swift [hunk 3]
```diff
@@ -254,11 +254,16 @@ public final class Collection : CollectionChangeObservable, Indexable {
     ///
     /// If the collection is deleted or the database is closed, a warning message will be logged.
     public func addChangeListener(queue: DispatchQueue?,
-                                  listener: @escaping (CollectionChange) -> Void) -> ListenerToken {
+                                  listener: @escaping (CollectionChange) -> Void) -> ListenerToken? {
```
Comment (pasin) [inline Swift/Collection.swift:257]: The returned ListenerToken is non-null.

File: Swift/CollectionChangeObservable.swift [hunk 4]
```diff
@@ -24,13 +24,15 @@ import Foundation
 public protocol CollectionChangeObservable {
     /// Add a change listener to listen to change events occurring to any documents in the collection.
     /// To remove the listener, call remove() function on the returned listener token.
-    /// Throw Illegal State Exception or equivalent if the default collection doesn’t exist.
-    func addChangeListener(listener: @escaping (CollectionChange) -> Void) -> ListenerToken
+    ///
+    /// If the collection is deleted or the database is closed, a warning message will be logged.
+    func addChangeListener(listener: @escaping (CollectionChange) -> Void) -> ListenerToken?
```
Comment (pasin) [inline Swift/CollectionChangeObservable.swift:29]: The returned ListenerToken is non-null.

File: Swift/CollectionChangeObservable.swift [hunk 5]
```diff
@@ -24,13 +24,15 @@ import Foundation
...
     
     /// Add a change listener to listen to change events occurring to any documents in the collection.
     /// If a dispatch queue is given, the events will be posted on the dispatch queue.
     /// To remove the listener, call remove() function on the returned listener token.
-    /// Throw Illegal State Exception or equivalent if the default collection doesn’t exist.
+    ///
+    /// If the collection is deleted or the database is closed, a warning message will be logged.
     func addChangeListener(queue: DispatchQueue?,
-                           listener: @escaping (CollectionChange) -> Void) -> ListenerToken
+                           listener: @escaping (CollectionChange) -> Void) -> ListenerToken?
```
Comment (pasin) [inline Swift/CollectionChangeObservable.swift:37]: The returned ListenerToken is non-null.

**Review by pasin — CHANGES_REQUESTED** [review]

File: Objective-C/CBLCollection.mm [hunk 6]
```diff
@@ -52,6 +58,9 @@ - (instancetype) initWithDB: (CBLDatabase*)db
         _name = slice2string(spec.name);
         _scope = [[CBLScope alloc] initWithDB: db name: slice2string(spec.scope)];
         
+        NSString* qName = $sprintf(@"Collection <%@>", _name);
```
Comment (pasin) [inline Objective-C/CBLCollection.mm:61]: Will it be OK if I have two instances of the same collection?

**Review by pasin — CHANGES_REQUESTED** [review]

File: Objective-C/CBLCollection.mm (hunk 6 above)
Comment (jayahariv) [reply Objective-C/CBLCollection.mm:61]: yes. I will update to include the self address as well. 

I copied this code from database, which means the same issue(non unique GCD queue can be created) can happen in Database as well .

File: Swift/Collection.swift (hunk 2 above)
Comment (pasin) [reply Swift/Collection.swift:247]: It still returns the ListenerToken but there will be no notifications. It's the same behavior when holding the ListenerToken but the database is closed later.

File: Objective-C/CBLCollection.mm (hunk 1 above)
Comment (jayahariv) [reply Objective-C/CBLCollection.mm:318]: right, there shouldn't be. we don't work with c4collection here.

**Review by pasin — APPROVED** [review]

**Pull request merged**